### PR TITLE
Live Chat

### PR DESCRIPTION
**Pull request by nick-funk**

### CODE REVIEW

File: src/core/client/stream/tabs/Live/AfterComments.tsx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+      { count, cursor: paginationCursor = cursor },
+      fragmentVariables
+    ) {
+      return {
+        count,
+        cursor: paginationCursor,
+        includeBefore: true,
+        includeAfter: true,
+        storyID: story.id,
+        flattenReplies: true,
```
Comment (alinhle) [inline src/core/client/stream/tabs/Live/AfterComments.tsx]: Some of these are not used?

File: src/core/client/stream/tabs/Live/LiveStream.tsx [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+      relay.hasMore() &&
+      !isLoadingMore &&
+      scrollEnabled
+    ) {
+      try {
+        setScrollEnabled(false);
+        await loadMore();
+
+        if (beforeComments.length > 0) {
+          const id = `comment-${beforeComments[0].__id}`;
```
Comment (alinhle) [inline src/core/client/stream/tabs/Live/LiveStream.tsx]: Request `id` explicitly and use that.

File: src/core/client/stream/tabs/Live/LiveStream.tsx [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+  const endRef = useRef(null);
+  useEffect(() => {
+    if (!endRef.current) {
+      return;
+    }
+
+    const end = endRef.current as any;
+    end.scrollIntoView();
+
+    setTimeout(() => {
```
Comment (alinhle) [inline src/core/client/stream/tabs/Live/LiveStream.tsx]: clear timeout

File: src/core/server/graph/schema/schema.graphql [hunk 4]
```diff
@@ -531,6 +531,11 @@ enum FEATURE_FLAG {
   overrides the configuration option available in the Admin UI for that tenant.
   """
   DISABLE_LIVE_UPDATES
+
+  """
+  Enables the chat UI for displaying a new Chat story mode.
+  """
+  CHAT
```
Comment (alinhle) [inline src/core/server/graph/schema/schema.graphql:543]: ENABLE_CHAT?

**Review by alinhle — CHANGES_REQUESTED** [review]

File: src/core/client/stream/tabs/Live/LiveChatContainer.tsx [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+      </IntersectionProvider>
+
+      {newReplyID && (
+        <div className={styles.scrollToNewReply}>
+          <Button onClick={scrollToNewReply} color="primary">
+            Reply posted below <Icon>arrow_downward</Icon>
+          </Button>
+        </div>
+      )}
+      {replyVisible && focusedComment && storyID && viewer && (
```
Comment (alinhle) [inline src/core/client/stream/tabs/Live/LiveChatContainer.tsx]:
```suggestion
      {replyVisible && focusedComment && viewer && (
```

File: src/core/client/stream/tabs/Live/LiveCreateCommentMutation.tsx [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+
+function sharedUpdater(
+  environment: Environment,
+  store: RecordSourceSelectorProxy,
+  input: LiveCreateCommentInput
+) {
+  const commentEdge = store
+    .getRootField("createComment")!
+    .getLinkedRecord("edge")!;
+
```
Comment (cvle) [inline src/core/client/stream/tabs/Live/LiveCreateCommentMutation.tsx:47]: What is this for?

File: src/core/client/stream/tabs/Live/LiveChatContainer.tsx [hunk 7]
```diff
@@ -0,0 +1,703 @@
...
+      return;
+    }
+
+    setCursor(newComment.cursor);
+
+    if (scrollToCommentTimeout.current) {
+      window.clearTimeout(scrollToCommentTimeout.current);
+    }
+
+    window.setTimeout(scrollToComment, 300);
```
Comment (alinhle) [inline src/core/client/stream/tabs/Live/LiveChatContainer.tsx]: I think you forgot to set the timeout ref:
`scrollToCommentTimeout.current = window.clearTimeout(scrollToCommentTimeout.current);`

**Comment by alinhle** [comment]
Noting this here so we don't forget:
- When story is closed and comments are empty, we would still encourage user to write comments even when it's not possible.
image

File: src/core/client/stream/tabs/Live/LiveTabQuery.tsx [hunk 8]
```diff
@@ -67,6 +68,14 @@ const LiveTabQuery: FunctionComponent = () => {
     [setCursor]
   );
 
+  const heartbeatIntervalMs = 1000;
+  const reloadOnIdle = useCallback((timePassedSeconds: number) => {
+    if (timePassedSeconds > 3 * heartbeatIntervalMs) {
+      window.location.reload();
+    }
+  }, []);
+  useHeartbeat(reloadOnIdle, heartbeatIntervalMs);
```
Comment (alinhle) [inline src/core/client/stream/tabs/Live/LiveTabQuery.tsx]: This looks a bit hardcore 👀

File: src/core/client/stream/tabs/Live/LiveCreateCommentMutation.tsx [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+  commentEdge.setValue(new Date().toISOString(), "cursor");
```
Comment (alinhle) [inline src/core/client/stream/tabs/Live/LiveCreateCommentMutation.tsx]: `commentEdge.setValue(new Date().toISOString(), "cursor");`

**Review by edoarso — APPROVED** [review]

**Comment by nick-funk** [comment]
Closing this PR because we have created an epic branch for live chat here:
https://github.com/coralproject/talk/pull/3504

**Pull request closed without merging**